New TF model inputs

jplu opened the pull request:
# What does this PR do?

This PR improves the way the inputs are handled in the TensorFlow models:

- It should be easier now to write a TensorFlow model
- The order of the inputs are now better handled, mostly when using Keras Tensors
- Replace all the occurrences of `inputs` by `input_ids` to make easier to understand what this parameter is for and to align with PyTorch input names and the Tokenizers outputs.

@LysandreJik @sgugger @patrickvonplaten let me know what you think about this new input processing, it is not finished yet but any comment will be helpful for me :)

sgugger commented on src/transformers/models/albert/modeling_tf_albert.py:
```diff
@@ -526,56 +527,65 @@ def call(
+                FutureWarning,
+            )
+            input_ids = kwargs.pop("inputs")
+
```
If the `kwargs` are not used elsewhere, adding a sanity check on whether they are an empty dict or not here would be great, something like:
```
if len(kwargs) > 0:
    raise ValueError(f"The following keyword arguments are not supported by this model: {list(kwargs.keys())}.")
```

sgugger commented on src/transformers/models/albert/modeling_tf_albert.py:
```diff
@@ -526,56 +527,65 @@ def call(
+            training=training,
+        )
+
+        if "inputs" in inputs:
+            warnings.warn(
+                "The `inputs` argument is deprecated and will be removed in a future version, use `input_ids` instead.",
+                FutureWarning,
+            )
+            inputs["input_ids"] = inputs.pop("inputs")
```
If this is globally done on all models, should this be in the `input_processing` function instead? It would avoid us having it in all models. Or is it just targeted on some specific models?
We could also pass the kwargs to the function `input_processing` and have the lines about `inputs` at 532 be treated inside that function too.

sgugger reviewed (commented):
Not sure if everything should be reviewed in detail as this is marked as draft. For now, I've looked at the function that prepares the inputs and one model.

I like this new functionality a lot and love the simplifications it brings in the modeling code for all the TF models.

One generic question: can we do all check about the `inputs` argument and deprecation warning in this `input_processing` function? This would be great as it would simplify even more the code of all models.

One small nit is that we're trying to consistently use f-string now in the lib, I made a comment on the first `ValueError` you have, but it applies for all others.

jplu commented:
@sgugger Thanks for these useful comments, I should have addressed them as you proposed. All the modification have been directly integrated in the new `input_processing` function. can you please also check the `modeling_tf_bart.py`, I have updated the documentation, let me know if I did something wrong.

@patrickvonplaten can you please check the same BART file as I have done some updates in order to make it able to properly handle the `return_dict=True`  and `return_dict=False` my updates don't affect the usual behaviour of the model even if the tests are ok.

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -236,6 +247,108 @@ def compute_loss(self, labels, logits):
+        )
+
+    for k, v in kwargs.items():
+        if isinstance(v, (tf.Tensor, bool, ModelOutput, tuple, list, dict)) or v is None:
```
can't we have an input of `int`? Reformer and RAG use ints as inputs to the forward (don't exist yet in TF, but might soon)

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -236,6 +247,108 @@ def compute_loss(self, labels, logits):
 
+def input_processing(func, input_ids, **kwargs):
+    signature = dict(inspect.signature(func).parameters)
+    _ = signature.pop("kwargs", None)
```
(nit) no need for `_ =`

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -236,6 +247,108 @@ def compute_loss(self, labels, logits):
+                f"Data of type {type(input_ids)} is not allowed only tf.Tensor is accepted for {parameter_names[0]}."
+            )
+
+    for name in parameter_names:
```
I wonder whether those three lines are needed...I think the default values of functions are always `None`, so does it make sense to set this value to `None` then?

jplu replied:
Yes it has, because they are not always filled by their default value (when creating a saved model for example), and then output might not contain all the parameters. So for such cases I prefer to be sure that all the needed parameters are properly filled.

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -236,6 +247,108 @@ def compute_loss(self, labels, logits):
+
+        del output["args"]
+
+    if "kwargs" in output:
```
Out of curiosity. What would be a case where it Is possible for `kwargs` to be in output at this point?

jplu replied:
When creating a saved model.

patrickvonplaten commented on src/transformers/models/bart/modeling_tf_bart.py:
```diff
@@ -1018,8 +1038,7 @@ def get_output_embeddings(self):
 )
-class TFBartForConditionalGeneration(TFPretrainedBartModel):
-    base_model_prefix = "model"
+class TFBartForConditionalGeneration(TFPretrainedBartModel, TFCausalLanguageModelingLoss):
```
Can we please not change the loss computation in this PR. `TFCausalLanguageModelingLoss` is different from Bart's compute loss function because Bart's loss function explicitly uses the pad_token_id as the ignore token and not -100. Sam did some special Bart things here that I'd like to take a look at in another PR.

jplu replied:
Ok, I put back the way as it was to compute the loss.

patrickvonplaten reviewed (commented):
I only looked at the `input_processing` function and TFBart. Overall, I think it looks great! I'm definitely in favor of this simplified API! 

One important thing however would be to not do any other major refactorings in TFBart. We should not refactor/change the loss computation of TFBart in this PR as mentioned in the comment earlier. I'm 100% on board with all the cosmetic changes `Tensor -> tf.Tensor` and the tf model outputs in Bart, but please no other big refactorings that are not related to the model outputs :-) 

Great work!

jplu commented:
@LysandreJik @sgugger @patrickvonplaten Do you think it would be interesting to have the same thing for the outputs? knowing that for graph mode compliance we will need to have such or such output behaviour depending the state we are (eager or not).

sgugger commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -236,6 +247,109 @@ def compute_loss(self, labels, logits):
+        if isinstance(v, (tf.Tensor, bool, int, ModelOutput, tuple, list, dict)) or v is None:
+            output[k] = v
+        else:
+            raise ValueError(f"Data of type {type(v)} is not allowed only tf.Tensor is accepted for {k}.")
```
Nit: we do accept more than tensors, so maybe put `(tf.Tensor, bool, int, ModelOutput, tuple, list, dict)` in a variable you reuse here to show all accepted types?

sgugger commented on src/transformers/models/electra/modeling_tf_electra.py:
```diff
@@ -1,4 +1,19 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" TF Elextra model. """
```
```suggestion
""" TF Electra model. """
```

sgugger commented on tests/test_modeling_tf_bart.py:
```diff
@@ -120,37 +118,6 @@ def test_inputs_embeds(self):
         # inputs_embeds not supported
         pass
 
-    def test_compile_tf_model(self):
```
Is this test removed because it's in the common tests?

jplu replied:
Exactly! I think Sam did rewrite it because of the different outputs, but basically there was no diff between the two (this one and the common).

sgugger commented on tests/test_modeling_tf_common.py:
```diff
@@ -753,7 +753,7 @@ def test_resize_token_embeddings(self):
     def test_lm_head_model_random_no_beam_search_generate(self):
         config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
-        input_ids = inputs_dict["input_ids"] if "input_ids" in inputs_dict else inputs_dict["inputs"]
+        input_ids = inputs_dict["input_ids"]  # if "input_ids" in inputs_dict else inputs_dict["inputs"]
```
I think we can clean up this comment and remove it thanks to all your hard work!

sgugger reviewed (approved):
Thanks a lot for this, the result is much much easier to read and clearer!
Remember to do separate PRs for separate issues please, a lot of changes in TFBart here are unrelated to the main focus of the PR.

jplu commented:
> Remember to do separate PRs for separate issues please, a lot of changes in TFBart here are unrelated to the main focus of the PR.

Do you prefer that I move the Bart changes into another PR and keep only the inputs changes? I don't mind to do this :)

jplu commented:
I have also updated the TF template.

sgugger commented:
> Do you prefer that I move the Bart changes into another PR and keep only the inputs changes? I don't mind to do this :)

Since @patrickvonplaten approved, I think they're okay here for this time (unless he says otherwise ;-) )

patrickvonplaten commented on src/transformers/models/bart/modeling_tf_bart.py:
```diff
@@ -916,93 +936,92 @@ def call(
-                input_ids=input_ids,
-                attention_mask=attention_mask,
+        # Replace by proper TF asserts
+        # assert (
```
can we maybe uncomment those lines or replace them with tf.assert

jplu replied:
This assert is not anymore useful because it was needed to be sure that the encoder output format is a `TFBaseModelOutput` now, thanks to my updates it can also be a Tuple, so I removed the comment.

patrickvonplaten reviewed (commented):
TFT5 and T5Bart look good to me! Just some commented-out lines that we can either uncomment for now or replace with `tf.debugging.assert`?

LysandreJik reviewed (approved):
I thinks this is a very welcome change, which clarifies the code. I would prefer if we didn't include this in the v4.0.0, however, as I think a bit of usage would be nice to ensure everything works as expected. We can still merge it once @patrickvonplaten and @sgugger's comments have been answered.

jplu commented:
I would prefer the same, it looks a bit too big to be added at this point of the release. I should have addressed all the comments. Let me know if I missed some.

LysandreJik reviewed (approved):
Thank you for your work on this very big PR.
